br/pkg/lightning/restore/data_sample_check.go

```diff
@@ -0,0 +1,308 @@
+
+	checkCfg *config.CheckOnly
+	wg       sync.WaitGroup
+
+	totalRows                    atomic.Int64
+	totalColumnCountMismatchRows atomic.Int64
+	totalInvalidCharRows         atomic.Int64
+}
+
+func newDataSampleCheck(controller *Controller) *dataSampleCheck {
+	return &dataSampleCheck{
+		controller: controller,
+		checkCfg:   controller.cfg.CheckOnly,
+	}
+}
+
+func (d *dataSampleCheck) logCheckFailedRow(row []types.Datum, fileMeta mydump.SourceFileMeta, offset int64, errMsg string) {
+	var sb strings.Builder
+	for i, d := range row {
+		if i > 0 {
+			sb.WriteString(", ")
+		}
+		// see Datum.String()
+		v := d.GetValue()
+		if b, ok := v.([]byte); ok && d.Kind() == types.KindBytes {
+			v = string(b)
+		}
+		sb.WriteString(fmt.Sprintf("%v", v))
+	}
+	log.L().Error("data check failed", zap.String("raw-data", sb.String()),
+		zap.String("file", fileMeta.Path),
+		zap.Int64("offset", offset),
+		zap.String("error", errMsg))
+}
+
+func (d *dataSampleCheck) checkRoutine(ctx context.Context, fileChan chan *sampledDataFileInfo, errChan chan error) {
+	defer d.wg.Done()
+
+	var resultErr error
+	defer func() {
+		errChan <- errors.Trace(resultErr)
+	}()
+
+	rc := d.controller
+	for fileInfo := range fileChan {
+		fileMeta := fileInfo.DataFile.FileMeta
+		fileParser, err := rc.createDataFileParser(ctx, fileMeta, config.DefaultCSVDataInvalidCharReplace)
+		if err != nil {
+			if errors.Cause(err) != context.Canceled {
+				resultErr = err
+			}
+			return
+		}
+
+		ignoreColumns, err := rc.cfg.Mydumper.IgnoreColumns.GetIgnoreColumns(fileInfo.Table.DB, fileInfo.Table.Name, rc.cfg.Mydumper.CaseSensitive)
+		if err != nil {
+			resultErr = err
+			return
+		}
+		ignoreColumnsMap := ignoreColumns.ColumnsMap()
+
+		ddlColumnMap := make(map[string]*model.ColumnInfo, len(fileInfo.TableInfo.Columns))
+		for _, col := range fileInfo.TableInfo.Columns {
+			ddlColumnMap[col.Name.L] = col
+		}
+		var columnNames []string
+		var rowsChecked, columnCountMismatchRows, invalidCharRows, lastOffset int64
+		columnCount := -1
+		var count int64
+		for {
+			count++
+			if d.checkCfg.Rows != config.CheckAllRows && count > d.checkCfg.Rows {
+				break
+			}
+			err = fileParser.ReadRow()
+			if err != nil {
+				if errors.Cause(err) == io.EOF {
+					break
+				}
+
+				if errors.Cause(err) != context.Canceled {
+					resultErr = err
+				}
+				fileParser.Close()
+				return
+			}
+
+			row := fileParser.LastRow()
+			if columnCount == -1 {
+				columnNames = fileParser.Columns()
+				if len(columnNames) == 0 {
+					columnCount = 0
+					columnNames = make([]string, 0, columnCount)
+					for _, col := range fileInfo.TableInfo.Columns {
+						if !col.Hidden {
+							columnNames = append(columnNames, col.Name.L)
+							columnCount++
+						}
+					}
+				} else {
+					columnCount = len(columnNames)
+				}
+			}
+			rowsChecked++
+			if len(row.Row) != columnCount {
+				columnCountMismatchRows++
+				d.logCheckFailedRow(row.Row, fileMeta, lastOffset,
```

Comment:
In most scenarios, column count mismatch implies all rows count are mismatch. This may output a lot of error logs which is hard to read. I think we should consider merging multiple logs or set a threshold.

Reply:
```
Besides, if there are many tables and one of the table has more files than others, this check may skip most of the tables. However, sampling more tables seems more better than sampling more files. I suggest sampling every table and perform random file choosing every table.
```
I think this is a scenario that needed to be considered. But we can noted it down first and gather some other feedbacks so we can improve it later.

Reply:
```
In most scenarios, column count mismatch implies all rows count are mismatch. This may output a lot of error logs which is hard to read. I think we should consider merging multiple logs or set a threshold.
```
Make sense. Maybe there may be multiple features could cause huge error messages. I think the better way is to add a common log security  component and not only threshold.  So I suggest we discuss it in a new task.